File: src/java/quarkus/src/main/java/com/benchmark/api/service/DatabaseService.java
package com.benchmark.api.service;

import com.benchmark.api.model.ComplexOrderResult;
import com.benchmark.api.model.User;
import io.smallrye.mutiny.Uni;
import io.vertx.mutiny.pgclient.PgPool;
import io.vertx.mutiny.sqlclient.Row;
import io.vertx.mutiny.sqlclient.RowSet;
import io.vertx.mutiny.sqlclient.Tuple;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;

@ApplicationScoped
public class DatabaseService {

    private static final Logger LOG = LoggerFactory.getLogger(DatabaseService.class);

    @Inject
    PgPool client;

    public Uni<User> findUserById(Integer id) {
        String query = "SELECT id, email, first_name, last_name, age, created_at FROM users WHERE id = $1";

        return client.preparedQuery(query)
                .execute(Tuple.of(id))
                .map(RowSet::iterator)
                .map(iterator -> {
                    if (iterator.hasNext()) {
                        Row row = iterator.next();
                        return mapUser(row);
                    }
                    return null;
                })
                .onFailure().recoverWithItem(t -> {
                    LOG.error("Error fetching user: {}", t.getMessage());
                    return null;
                });
    }

    public Uni<List<ComplexOrderResult>> findComplexOrders(Integer days) {
        String query = """
            -- Normative SQL, see contracts/rest/canonical-payloads.md. The previous
            -- query joined order_items, aggregated quantity*price and ordered without
            -- a tiebreak, so it ran a heavier query than the other implementations and
            -- its rows came back in arbitrary order among equal values.
            SELECT
                u.id AS "userId",
                u.first_name || ' ' || u.last_name AS "userName",
                COUNT(o.id) AS "totalOrders",
                COALESCE(SUM(o.total_amount), 0) AS "totalValue",
                COALESCE(AVG(o.total_amount), 0) AS "averageOrderValue"
            FROM users u
            INNER JOIN orders o ON u.id = o.user_id
                WHERE o.created_at >= NOW() - INTERVAL '1 day' * $1
            GROUP BY u.id, u.first_name, u.last_name
            ORDER BY "totalOrders" DESC, u.id
            LIMIT 100
            """;

        return client.preparedQuery(query)
                .execute(Tuple.of(days))
                .map(this::mapComplexResults)
                .onFailure().recoverWithItem(t -> {
                    LOG.error("Error fetching complex orders: {}", t.getMessage());
                    return new ArrayList<>();
                });
    }

    public Uni<Boolean> healthCheck() {
        return client.query("SELECT 1")
                .execute()
                .map(rows -> rows.size() > 0)
                .onFailure().recoverWithItem(false);
    }

    private User mapUser(Row row) {
        return new User(
                row.getInteger("id"),
                row.getString("email"),
                row.getString("first_name"),
                row.getString("last_name"),
                row.getInteger("age"),
                row.getOffsetDateTime("created_at").toString()
        );
    }

    private List<ComplexOrderResult> mapComplexResults(RowSet<Row> rows) {
        List<ComplexOrderResult> results = new ArrayList<>();
        for (Row row : rows) {
            results.add(new ComplexOrderResult(
                    row.getInteger("userId"),
                    row.getString("userName"),
                    row.getLong("totalOrders"),
                    row.getDouble("totalValue"),
                    row.getDouble("averageOrderValue")
            ));
        }
        return results;
    }
}

File: src/csharp/FastEndpoints/Services/DatabaseService.cs
using BenchmarkFastEndpoints.Models;
using Npgsql;

namespace BenchmarkFastEndpoints.Services;

public interface IDatabaseService
{
    Task<User?> GetUserByIdAsync(int id);
    Task<ComplexResult[]> GetComplexQueryAsync(int days);
}

public class DatabaseService : IDatabaseService
{
    private readonly string _connectionString;

    public DatabaseService(IConfiguration configuration)
    {
        var url = Environment.GetEnvironmentVariable("DATABASE_URL")
            ?? configuration.GetConnectionString("DefaultConnection")
            ?? throw new InvalidOperationException("Database connection string not found");

        // Parse postgresql://user:password@host:port/database
        // Handle @ in password by splitting from last @
        if (url.StartsWith("postgresql://"))
        {
            var lastAt = url.LastIndexOf('@');
            var schemeEnd = url.IndexOf("://");
            var userPass = url.Substring(schemeEnd + 3, lastAt - schemeEnd - 3);
            var hostPortDb = url.Substring(lastAt + 1);
            var user = userPass.Split(':')[0];
            var password = userPass.Contains(':') ? userPass.Substring(userPass.IndexOf(':') + 1) : "";
            var host = hostPortDb.Split(':')[0];
            var portDb = hostPortDb.Substring(hostPortDb.IndexOf(':') + 1);
            var port = portDb.Split('/')[0];
            var database = portDb.Contains('/') ? portDb.Substring(portDb.IndexOf('/') + 1) : "";

            _connectionString = $"Host={host};Port={port};Database={database};Username={user};Password={password};Maximum Pool Size=25;Connection Timeout=30";
        }
        else
        {
            _connectionString = url;
        }
    }

    public async Task<User?> GetUserByIdAsync(int id)
    {
        using var connection = new NpgsqlConnection(_connectionString);
        await connection.OpenAsync();

        using var cmd = new NpgsqlCommand(
            "SELECT id, email, first_name, last_name, age, created_at FROM users WHERE id = @Id",
            connection);
        cmd.Parameters.AddWithValue("Id", id);

        using var reader = await cmd.ExecuteReaderAsync();
        if (await reader.ReadAsync())
        {
            return new User
            {
                Id = reader.GetInt32(0),
                Email = reader.GetString(1),
                FirstName = reader.GetString(2),
                LastName = reader.GetString(3),
                Age = reader.IsDBNull(4) ? null : reader.GetInt32(4),
                CreatedAt = reader.GetDateTime(5)
            };
        }

        return null;
    }

    public async Task<ComplexResult[]> GetComplexQueryAsync(int days)
    {
        using var connection = new NpgsqlConnection(_connectionString);
        await connection.OpenAsync();

        var results = new List<ComplexResult>();

        // Normative SQL, see contracts/rest/canonical-payloads.md. The previous
        // query joined order_items and aggregated quantity*price -- a
        // materially heavier query than the one other implementations ran --
        // and ordered only by TotalValue, so rows with equal values came back
        // in arbitrary order and the response was not reproducible.
        using var cmd = new NpgsqlCommand(@"
            SELECT
                u.id AS UserId,
                u.first_name || ' ' || u.last_name AS UserName,
                COUNT(o.id) AS TotalOrders,
                COALESCE(SUM(o.total_amount), 0) AS TotalValue,
                COALESCE(AVG(o.total_amount), 0) AS AverageOrderValue
            FROM users u
            INNER JOIN orders o ON u.id = o.user_id
                WHERE o.created_at >= NOW() - make_interval(days => @Days)
            GROUP BY u.id, u.first_name, u.last_name
            ORDER BY TotalOrders DESC, u.id
            LIMIT 100", connection);

        cmd.Parameters.AddWithValue("Days", days);

        using var reader = await cmd.ExecuteReaderAsync();
        while (await reader.ReadAsync())
        {
            results.Add(new ComplexResult
            {
                UserId = reader.GetInt32(0),
                UserName = reader.GetString(1),
                TotalOrders = reader.GetInt64(2),
                TotalValue = reader.GetDecimal(3),
                AverageOrderValue = reader.GetDecimal(4)
            });
        }

        return results.ToArray();
    }
}
